### PR TITLE
Support Independent(Normal) observations in GaussianHMM

### PR DESCRIPTION
**Pull request by fritzo**
This adds support for diagonal observation covariances in `GaussianHMM`.

My use case is an HMM with very large observation space, say latent dimension of M=32 but observed dimension of N=2000. This PR reduces complexity from `M^3+N^3` to `M^3+MN`.

## Tested
- unit test for `matrix_and_mvn_to_gaussian()` with `AffineNormal`
- added diag versions of shape tests to `GaussianHMM` tests

### CODE REVIEW

File: pyro/ops/gaussian.py [hunk 1]
```diff
@@ -207,6 +207,49 @@ def event_logsumexp(self):
                 chol_P.diagonal(dim1=-2, dim2=-1).log().sum(-1))
 
 
+class AffineNormal:
```
Comment (fehiepsi) [inline pyro/ops/gaussian.py]: nit: do we need to subclass `object` here?

File: pyro/ops/gaussian.py [hunk 2]
```diff
@@ -207,6 +207,49 @@ def event_logsumexp(self):
...
+    """
+    Represents a conditional diagonal normal distribution over a random
+    variable ``Y`` whose mean is an affine function of a random variable ``X``.
+    The likelihood of ``X`` is thus::
+
+        LinearNormal(matrix, loc, scale).condition(y).log_density(x)
```
Comment (fehiepsi) [inline pyro/ops/gaussian.py]: nit: LinearNormal -> AffineNormal
You might want to use the name `MultivariateNormalDiag` as in tensorflow (though I don't like this idea much).

File: pyro/ops/gaussian.py (hunk 1 above)
Comment (fritzo) [reply pyro/ops/gaussian.py]: Python 3 no longer requires `(object)`. I'll also remove `(object)` from `Gaussian` in this file, but let's do general cleanup in a separate PR.

File: pyro/ops/gaussian.py (hunk 2 above)
Comment (fritzo) [reply pyro/ops/gaussian.py]: Done.

I agree that `AffineNormal` is a weird name, but I don't think `MultivariateNormalDiag` is the right name either. This class really serves a very narrow role as a likelihood of a diag normal observation of an affine transformation. I'm open to renaming / refactoring once we see whether and how this pattern generalizes.

**Comment by fritzo** [comment]
Thanks for reviewing, @fehiepsi !

**Pull request merged**